sw/device/silicon_creator/mask_rom/mask_rom.c

```diff
@@ -15,17 +15,48 @@
 #include "sw/device/lib/runtime/hart.h"
 #include "sw/device/lib/runtime/print.h"
 #include "sw/device/silicon_creator/lib/drivers/keymgr.h"
+#include "sw/device/silicon_creator/lib/drivers/lifecycle.h"
 #include "sw/device/silicon_creator/lib/drivers/uart.h"
+#include "sw/device/silicon_creator/lib/error.h"
+#include "sw/device/silicon_creator/lib/shutdown.h"
 #include "sw/device/silicon_creator/lib/sigverify.h"
 #include "sw/device/silicon_creator/mask_rom/romextimage.h"
 #include "sw/device/silicon_creator/mask_rom/sigverify_keys.h"
 
 #include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
 
-void mask_rom_exception_handler(void) { wait_for_interrupt(); }
-void mask_rom_nmi_handler(void) { wait_for_interrupt(); }
+static inline rom_error_t mask_rom_irq_error(void) {
+  uint32_t mcause;
+  CSR_READ(CSR_REG_MCAUSE, &mcause);
+  // Shuffle the mcause bits into the uppermost byte of the word and report
+  // the cause as kErrorInterrupt.
+  // Based on the ibex verilog, it appears that the most significatn bit
+  // indicates whether the cause is an exception (0) or external interrupt (1),
+  // and the 5 least significant bits indicate which exception/interrupt.
+  //
+  // Preserve the MSB and shift the 5 LSBs into the upper byte.
+  mcause = (mcause & 0x80000000) | ((mcause & 0x1f) << 24);
```

Comment:
I've added the implementation with the bitfield library.
It is, IMNSHO, worse for readability than the single expression with mcause.

Reply:
For single-bit bit fields you can use `bitfield_bit32_{read,write}`. I think it would also be fine to do the masking of the reason as part of the write operation. Something like:

```
rom_error_t error = kErrorInterrupt;
error = bitfield_bit32_write(error, 31, bitfield_bit32_read(mcause, 31));
error = bitfield_field32_write(error, (bitfield_field32_t){.index = 24, .mask = 0x7f}, mcause);
return error;
```

If you prefer the original code (doing direct bit manipulation) though I think that is fine too. The bitfield operations tend to make more sense when the fields are defined elsewhere. For example, it might make sense to turn mcause into an 8-bit value using direct bit manipulation but then set the relevant 8-bit field in the error code using the bitfield library.